lib_nn/api/geom/ImageGeometry.hpp

```diff
@@ -0,0 +1,236 @@
+   * in padding (i.e. beyond the bounds of the image).
+   */
+  int Index(const int row, const int col, const int channel) const;
+
+  /**
+   * Get the buffer index associated with the specified image coordinates,
+   * assuming the image is backed by a linear buffer.
+   */
+  int Index(const ImageVect &input_coords) const;
+
+  /**
+   * Get the memory stride (in bytes) required to move by the specified amount
+   * within this geometry.
+   */
+  mem_stride_t GetStride(const int rows, const int cols, const int chans) const;
+
+  /**
+   * Get the memory stride (in bytes) required to move by the specified amount
+   * within this geometry.
+   */
+  mem_stride_t GetStride(const ImageVect &vect) const;
+
+  /**
+   * Get the memory stride (in bytes) required to move between two locations
+   * within this geometry.
+   */
+  mem_stride_t GetStride(const ImageVect &from, const ImageVect &to) const;
+
+  /**
+   * Check whether the specified coordinates refer to an element within this
+   * image geometry.
+   */
+  bool IsWithinImage(const ImageVect &coords) const;
+
+  /**
+   * Check whether the specified coordinates refer to an element within this
+   * image geometry.
+   */
+  bool IsWithinImage(const int row, const int col, const int channel) const;
+
+  /**
+   * Get a reference to the specified element from the provided image.
+   * An assertion is made that the specified location is within the geometry's
+   * bounds.
+   */
+  template <typename T>
+  T &Element(T *img_base, const int row, const int col,
+             const int channel) const;
+
+  /**
+   * Get the value of the specified element from the provided image.
+   * If the coordinates refer to a location outside the image geometry pad_value
+   * is returned.
+   */
+  template <typename T>
+  T Get(const T *img_base, const ImageVect coords, const T pad_value = 0) const;
+
+  /**
+   * Get the value of the specified element from the provided image.
+   * If the coordinates refer to a location outside the image geometry pad_value
+   * is returned.
+   */
+  template <typename T>
+  T Get(const T *img_base, const int row, const int col, const int channel,
+        const T pad_value = 0) const;
+
+  /**
+   * Apply an operation per element of an image.
+   *
+   * void PerPixelOp(const int row,
+   *                 const int col,
+   *                 const int channel,
+   *                 T& value);
+   */
+  template <typename T>
+  void ApplyOperation(T *image_base,
+                      std::function<void(int, int, int, T &)> func) const;
+
+  /**
+   * Determine whether this ImageGeometry is equal to another.
+   */
+  bool operator==(ImageGeometry other) const;
+};
+
+/////////////////////////
+template <typename T>
+T &ImageGeometry::Element(T *img_base, const int row, const int col,
+                          const int channel) const {
+  assert(IsWithinImage(row, col, channel));
+  int index = row * this->width * this->depth + col * this->depth + channel;
```

Comment:
I'm not sure how the calculation is meant to be done. Is `this->height` missing here?

Reply:
nothing is missing here as the leading dimension is not required for dereferencing